### PR TITLE
Support opendnssec 2.1.6

### PR DESCRIPTION
**Pull request by flo-renaud**
### Support opendnssec 2.1.6
The installation of IPA DNS server is using ods-ksmutil, but
openddnssec 2.1.6 does not ship any more /usr/bin/ods-ksmutil. The tool
is replaced by /usr/sbin/ods-enforcer and /usr/sbin/ods-enforcer-db-setup.
    
The master branch currently supports fedora 30+, but fedora 30 and 31 are
still shipping opendnssec 1.4 while fedora 32+ is shipping opendnssec 2.1.6.
Because of this, the code needs to check at run-time if the ods-ksmutil
command is available. If the file is missing, the code falls back to
the new ods-enforcer and ods-enforcer-db-setup commands.

This commit defines paths.ODS_ENFORCER and paths.ODS_ENFORCER_DB_SETUP
for all platforms, but the commands are used only if ods-ksmutil is not found.
    
Fixes: https://pagure.io/freeipa/issue/8214

### Remove the <Interval> from opendnssec conf
    
In opendnssec 2.1.6, the <Interval> element is not supported in the
configuration file.

Note: This PR depends on https://src.fedoraproject.org/rpms/opendnssec/pull-request/2 because currently the opendnssec package does not create the /var/opendnssec/enforcer directory.
BZ https://bugzilla.redhat.com/show_bug.cgi?id=1809492 opened to track the issue.

### CODE REVIEW

File: daemons/dnssec/ipa-ods-exporter.in [hunk 1]
```diff
@@ -548,8 +568,18 @@ def send_systemd_reply(conn, reply):
         # This is necessary to let Enforcer to unlock the ODS DB.
         if six.PY3:
             reply = reply.encode('ascii')
-        conn.send(reply + b'\n')
-        conn.shutdown(socket.SHUT_RDWR)
+        if ods_version == 1:
```
Comment (abbra) [inline daemons/dnssec/ipa-ods-exporter.in]: `if ods_version() == 1:`

**Comment by abbra** [comment]
Few more TODO items from the investigation:

 - database format has changed between 1.4 and 2.x, needs to be handled in upgrade code
 - we should split opendnssec handling in `ipa-ods-exporter` into two separate modules, `opendnssec14` and `opendnssec2x`, implement differences there in a common interface and import proper one at runtime in `ipa-ods-exporter`

Database conversion is handled by https://github.com/opendnssec/opendnssec/tree/2.1/master/enforcer/utils/1.4-2.0_db_convert. This tool has been removed in the current git development branch, so we might want to incorporate its logic into our upgrade code directly.

**Comment by tiran** [comment]
Migration is handled by RPM package update code. Looks like Paul is shipping the migration script with the RPM, too. https://src.fedoraproject.org/rpms/opendnssec/blob/master/f/opendnssec.spec#_126

**Comment by abbra** [comment]
More issues:
 - `named-pkcs11` needs access to `/var/lib/ipa/dnssec/tokens/*` which are only available to `ods` group
 - if I remove and re-add DNS zone, an attempt to synchronize LDAP and local HSM store happens but does not process a response that a zone wasn't found in HSM. We get into a loop of trying to delete DNS zone from HSM infinitely. I think `ipa-dnskeysyncd` needs to recognize the case:
```
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipaserver.dnssec.odsmgr: DEBUG    LDAP zones: {'ba8a6809-607e-11ea-b0a9-bd2ea43c1cc3': <DNS name dnssec.test.>}
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipaserver.dnssec.keysyncer: INFO     Initial LDAP dump is done, sychronizing with ODS and BIND
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipapython.ipautil: DEBUG    Starting external process
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipapython.ipautil: DEBUG    args=['/usr/sbin/ods-enforcer', 'zonelist', 'export']
Mar 07 14:26:21 master.ipa.test ods-enforcerd[38540]: INFO: The XML in /etc/opendnssec/zonelist.xml.new is valid
Mar 07 14:26:21 master.ipa.test ods-enforcerd[38540]: [zonelist_export_cmd] zonelist exported to /etc/opendnssec/zonelist.xml successfully
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipapython.ipautil: DEBUG    Process finished, return code=0
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipapython.ipautil: DEBUG    stdout=Exported zonelist to /etc/opendnssec/zonelist.xml successfully
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipapython.ipautil: DEBUG    stderr=
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipaserver.dnssec.odsmgr: DEBUG    ODS zones: {'603bd801-607b-11ea-b0a9-bd2ea43c1cc3': <DNS name dnssec.test.>}
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipaserver.dnssec.odsmgr: INFO     Zones removed from LDAP: [('603bd801-607b-11ea-b0a9-bd2ea43c1cc3', <DNS name dnssec.test.>)]
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipaserver.dnssec.odsmgr: INFO     Zones added to LDAP: [('ba8a6809-607e-11ea-b0a9-bd2ea43c1cc3', <DNS name dnssec.test.>)]
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipapython.ipautil: DEBUG    Starting external process
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipapython.ipautil: DEBUG    args=['/usr/sbin/ods-enforcer', 'zone', 'delete', '--zone', 'dnssec.test']
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipapython.ipautil: DEBUG    Process finished, return code=1
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipapython.ipautil: DEBUG    stdout=
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipapython.ipautil: DEBUG    stderr=Unable to delete zone, zone dnssec.test not found!
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: Traceback (most recent call last):
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:   File "/usr/libexec/ipa/ipa-dnskeysyncd", line 117, in <module>
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:     while ldap_connection.syncrepl_poll(all=1, msgid=ldap_search):
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:   File "/usr/lib64/python3.8/site-packages/ldap/syncrepl.py", line 464, in syncrepl_poll
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:     self.syncrepl_refreshdone()
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:   File "/usr/lib/python3.8/site-packages/ipaserver/dnssec/keysyncer.py", line 125, in syncrepl_refreshdone
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:     self.ods_sync()
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:   File "/usr/lib/python3.8/site-packages/ipaserver/dnssec/keysyncer.py", line 183, in ods_sync
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:     self.odsmgr.sync()
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:   File "/usr/lib/python3.8/site-packages/ipaserver/dnssec/odsmgr.py", line 206, in sync
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:     self.del_ods_zone(name)
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:   File "/usr/lib/python3.8/site-packages/ipaserver/dnssec/odsmgr.py", line 174, in del_ods_zone
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:     output = self.ksmutil(cmd)
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:   File "/usr/lib/python3.8/site-packages/ipaserver/dnssec/odsmgr.py", line 139, in ksmutil
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:     result = tasks.run_ods_manager(params, capture_output=True)
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:   File "/usr/lib/python3.8/site-packages/ipaplatform/base/tasks.py", line 352, in run_ods_manager
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:     return ipautil.run(cmd, **kwargs)
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:   File "/usr/lib/python3.8/site-packages/ipapython/ipautil.py", line 597, in run
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]:     raise CalledProcessError(
Mar 07 14:26:21 master.ipa.test ipa-dnskeysyncd[38750]: ipapython.ipautil.CalledProcessError: CalledProcessError(Command ['/usr/sbin/ods-enforcer', 'zone', 'delete', '--zone', 'dnssec.test'] returned non-zero exit status 1: 'Unable to delete zone, zone dnssec.test not found!\n')
```

**Comment by abbra** [comment]
For the first issue logs are:
```
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: bug in zone_master_reconfigure(): default TTL cannot be changed at run-time
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: zone dnssec.test/IN (signed): loaded serial 1583590695
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: zone dnssec.test/IN (signed): receive_secure_serial: unchanged
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: zone dnssec.test/IN (signed): reconfiguring zone keys
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/be40e04d-f9b8-6037-304a-ca0219354d29.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/be40e04d-f9b8-6037-304a-ca0219354d29.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/a05f8c8d-26d1-895c-e2bb-c4902d7e5d7c.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/a05f8c8d-26d1-895c-e2bb-c4902d7e5d7c.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/10b561b8-ce59-0315-9d09-d1b5fa2793bd.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/10b561b8-ce59-0315-9d09-d1b5fa2793bd.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/31f45f85-3db6-b64e-d789-aeea1fbf6ed1.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/31f45f85-3db6-b64e-d789-aeea1fbf6ed1.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/71984fc7-efa1-0dd6-b43c-68d9aa61b837.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/71984fc7-efa1-0dd6-b43c-68d9aa61b837.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/a26b3c66-84b1-585b-5775-23b9f772fb5a.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/a26b3c66-84b1-585b-5775-23b9f772fb5a.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/af7cd23d-b678-f25d-4e92-ee3f0b4f0153.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/af7cd23d-b678-f25d-4e92-ee3f0b4f0153.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/b7111b48-de73-3d3e-5781-e6b15e24fb54.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/b7111b48-de73-3d3e-5781-e6b15e24fb54.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/c491c1df-dee6-0cb3-b57c-eba0a194ef83.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/c491c1df-dee6-0cb3-b57c-eba0a194ef83.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/d72061df-fd51-e519-016b-14c1e3bf0f13.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/d72061df-fd51-e519-016b-14c1e3bf0f13.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/f9260467-8e7a-add1-48b2-4044f26d9725.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/f9260467-8e7a-add1-48b2-4044f26d9725.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/3d0f55b2-76a3-74fa-e883-510e271c5d60.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/3d0f55b2-76a3-74fa-e883-510e271c5d60.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/0d50c965-f85f-65eb-a519-02487ba43fed.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/0d50c965-f85f-65eb-a519-02487ba43fed.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/15fecd30-97cb-bde3-8e4c-a9f0f21793fc.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/15fecd30-97cb-bde3-8e4c-a9f0f21793fc.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/1f917b46-65a0-c1ec-7f1a-d7ac3e73132c.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/1f917b46-65a0-c1ec-7f1a-d7ac3e73132c.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/277f13fe-3f0c-4bfc-dd93-1da074c23361.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: File.cpp(94): Could not open the file (Permission denied): /var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/277f13fe-3f0c-4bfc-dd93-1da074c23361.object
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: dns_dnssec_findmatchingkeys: error reading key file Kdnssec.test.+008+54271.private: not found
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: zone dnssec.test/IN (signed): next key event: 07-Mar-2020 15:18:14.718
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: zone dnssec.test/IN (signed): could not get zone keys for secure dynamic update
Mar 07 14:18:14 master.ipa.test named-pkcs11[37128]: zone dnssec.test/IN (signed): serial 1583590696 (unsigned 1583590696)
```

From outside the container I am using to do this test, I can see what `named-pkcs11` is doing, with strace:
```
4062270 write(12, "07-Mar-2020 14:12:31.763 zone dnssec.test/IN (signed): reconfiguring zone keys\n", 79 <unfinished ...>
4062269 <... recvmsg resumed>{msg_namelen=128}, 0) = -1 EAGAIN (Resource temporarily unavailable)
4062272 <... epoll_ctl resumed>)        = 0
4062269 write(8, "\5\2\0\0\375\377\377\377", 8 <unfinished ...>
4062272 read(6,  <unfinished ...>
4062269 <... write resumed>)            = 8
4062272 <... read resumed>"\3\2\0\0\375\377\377\377", 8) = 8
4062270 <... write resumed>)            = 79
4062272 epoll_ctl(9, EPOLL_CTL_ADD, 515, {EPOLLIN, {u32=515, u64=515}} <unfinished ...>
4062269 futex(0x7f92ef2fc0cc, FUTEX_WAIT_PRIVATE, 0, NULL <unfinished ...>
4062272 <... epoll_ctl resumed>)        = 0
4062270 openat(AT_FDCWD, "dyndb-ldap/ipa/master/dnssec.test/keys/", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY <unfinished ...>
4062272 read(6, "\5\2\0\0\375\377\377\377", 8) = 8
4062272 epoll_ctl(9, EPOLL_CTL_ADD, 517, {EPOLLIN, {u32=517, u64=517}} <unfinished ...>
4062270 <... openat resumed>)           = 10
4062272 <... epoll_ctl resumed>)        = 0
4062270 fstat(10,  <unfinished ...>
4062272 read(6,  <unfinished ...>
4062270 <... fstat resumed>{st_mode=S_IFDIR|0770, st_size=170, ...}) = 0
4062272 <... read resumed>0x7f92ecac6d28, 8) = -1 EAGAIN (Resource temporarily unavailable)
4062270 mprotect(0x7f92e0064000, 24576, PROT_READ|PROT_WRITE <unfinished ...>
4062272 epoll_wait(9,  <unfinished ...>
4062270 <... mprotect resumed>)         = 0
4062270 getdents64(10, /* 6 entries */, 32768) = 248
4062270 openat(AT_FDCWD, "dyndb-ldap/ipa/master/dnssec.test/keys//Kdnssec.test.+008+54271.key", O_RDONLY) = 13
4062270 fstat(13, {st_mode=S_IFREG|0660, st_size=498, ...}) = 0
4062270 mprotect(0x7f92e006a000, 4096, PROT_READ|PROT_WRITE) = 0
4062270 read(13, "; This is a key-signing key, keyid 54271, for dnssec.test.\n; Created: 20200307141019 (Sat Mar  7 14:10:19 2020)\ndnssec.test. IN DNSKEY 257 3 8 AwEAAZw2hC/+2pNHgUyEiOF7brWWqDf649k6UsVEOB4t4aNK2zUFxnPK acMIHG1fyzPwpUm+O9ekEKNKeifJLUvpW/
wdRFqD8CJ++D9WmAzG/vHC KWVJB9/waRbE3R0Yv1ozNAMNj7eZJROWlYW88603xcKYw0ngDuqO0N56 YQlfejwyp7E9GJFtCbJs+PMgUHlSNUGW3qh8OX7g4EIwyj4hhLBqq9SF qcZb+9ppzpGrjMd4SFJPC4oYhcMorohS++CmWbPHeSMCyolH4C55ohE9 1YWoFc9rV+ingJb4DKzXJb/xMhpf356PhO3u2zkNg/Z8zyVYYjk2Fb/f IGw
3OhQXuYU=\n", 4096) = 498
4062270 close(13)                       = 0
4062270 openat(AT_FDCWD, "dyndb-ldap/ipa/master/dnssec.test/keys//Kdnssec.test.+008+54271.private", O_RDONLY) = 13
4062270 fstat(13, {st_mode=S_IFREG|0660, st_size=598, ...}) = 0
4062270 read(13, "Private-key-format: v1.3\nAlgorithm: 8 (RSASHA256)\nModulus: nDaEL/7ak0eBTISI4XtutZaoN/rj2TpSxUQ4Hi3ho0rbNQXGc8ppwwgcbV/LM/ClSb4716QQo0p6J8ktS+lb/B1EWoPwIn74P1aYDMb+8cIpZUkH3/BpFsTdHRi/WjM0Aw2Pt5klE5aVhbzzrTfFwpjDSeAO6o7Q3nphCV96PDKnsT0
YkW0Jsmz48yBQeVI1QZbeqHw5fuDgQjDKPiGEsGqr1IWpxlv72mnOkauMx3hIUk8LihiFwyiuiFL74KZZs8d5IwLKiUfgLnmiET3VhagVz2tX6KeAlvgMrNclv/EyGl/fno+E7e7bOQ2D9nzPJVhiOTYVv98gbDc6FBe5hQ==\nPublicExponent: AQAB\nEngine: cGtjczExAA==\nLabel: cGtjczExOm9iamVjdD0wM2M1MTAxOTdj
ZDg5MzlhYmU1ZmU4MGM0YWU5NjZmODtwaW4tc291cmNlPS92YXIvbGliL2lwYS9kbnNzZWMvc29mdGhzbV9waW4A\nCreated: 20200307141019\n", 4096) = 598
4062270 read(13, "", 4096)              = 0
4062270 openat(AT_FDCWD, "/var/lib/ipa/dnssec/softhsm_pin", O_RDONLY) = 14
4062270 fstat(14, {st_mode=S_IFREG|0770, st_size=30, ...}) = 0
4062270 mprotect(0x7f92e006b000, 4096, PROT_READ|PROT_WRITE) = 0
4062270 read(14, "2RaPlH0WRHsaKtlKoY7LjkqQzC8Qdb", 4096) = 30
4062270 read(14, "", 4096)              = 0
4062270 close(14)                       = 0
4062270 openat(AT_FDCWD, "/var/lib/ipa/dnssec/tokens/0772e0b5-2cf3-75d3-bad8-af7c07db8ff9/generation", O_RDONLY) = 14
4062270 fcntl(14, F_GETFL)              = 0x8000 (flags O_RDONLY|O_LARGEFILE)
4062270 fcntl(14, F_SETLKW, {l_type=F_RDLCK, l_whence=SEEK_SET, l_start=0, l_len=0}) = 0
4062270 fstat(14, {st_mode=S_IFREG|0660, st_size=8, ...}) = 0
```

**Comment by abbra** [comment]
Note that we do create `IPA_DNSSEC_DIR` with permissions to read it for `named` group but OpenDNSSEC utilities and daemons all run under `ods` user and explicitly set content in it (and subdirectories) to `ods:ods` because that's how the package is configured and how configuration file is set. Which means we should really add `named` to `ods` group rather than giving access to `named` group.

**Comment by abbra** [comment]
Correction: looks like we have already added fixing file and directory permissions in `ipaserver.dnssec.bindmgr`.

**Comment by abbra** [comment]
So, my understanding of the current state:

 - [x] database schema change happened in OpenDNSSEC 2.0, FreeIPA code needs changes (done)
 - [x] internal protocol between DNSSEC signer and a daemon which FreeIPA wraps with own code changed in OpenDNSSEC 2.0, FreeIPA code needs update (done)
 - [ ] ksm-util utility gone, its replacement got few commands changing its output, FreeIPA code needs update (partially done)
 - [x] DNSSEC default policy for a zone is not loaded when the database is initialized, FreeIPA code needs update (done)
 - [ ] removal of DNSSEC details in case DNS zone is removed triggers unhandled response from the new ksm-util replacement utility, FreeIPA code needs changes (not done yet)

File: ipaserver/dnssec/odsmgr.py [hunk 2]
```diff
@@ -159,10 +159,20 @@ def get_ods_zonelist(self):
...
+        name = name.relativize(dns.name.root)
         cmd = ['zone', 'add', '--zone', str(name), '--input', zone_path]
-        output = self.ksmutil(cmd)
-        logger.info('%s', output)
-        self.notify_enforcer()
+        output = None
+        try:
+            output = self.ksmutil(cmd)
+        except ipautil.CalledProcessError as e:
+            # Zone already doesn't exist in HSM
```
Comment (abbra) [inline ipaserver/dnssec/odsmgr.py]: Comment needs to be fixed

**Comment by flo-renaud** [comment]
The test failure in fedora-rawhide/test_dnssec is not linked to this PR but rather to the upgrade of tomcat to tomcat-1:9.0.31-1 (see the nightly tests [testing_master-rawhide] / PR #[152](https://github.com/freeipa-pr-ci2/freeipa/pull/152))

**Comment by abbra** [comment]
OpenDNSSEC setup failed on rawhide:
```
[ipatests.pytest_ipa.integration.host.Host.replica0.cmd41] Exit code: 1
ipa: ERROR: stderr: Unexpected error - see /var/log/ipaserver-install.log for details:
CalledProcessError: CalledProcessError(Command ['/usr/sbin/ods-enforcer', 'zonelist', 'export'] returned non-zero exit status 201: 'Unable to connect to engine. connect() failed: No such file or directory ("/var/run/opendnssec/enforcer.sock")\n')

```

**Comment by flo-renaud** [comment]
Tests were green except for pylint:
- azure: [Details](https://github.com/freeipa/freeipa/pull/4295/checks?check_run_id=503677414)
- fedora-latest/build: [Details](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/71e23c74-647d-11ea-8d16-fa163e07b2f0)
-  fedora-latest/test_dnssec: [Details](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/abea1b2a-647e-11ea-bfbb-fa163e0a82d3)
- fedora-rawhide/build: [Details](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/7faff328-647d-11ea-bee9-fa163e6233d3)
- fedora-rawhide/test_dnssec: [Details](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/44b58d1c-647f-11ea-8277-fa163e401a00)

I am removing the temp commit + fixing pylint in a single pass.

**Comment by abbra** [comment]
LGTM.

**Comment by flo-renaud** [comment]
master:

* 7ae1352c72b81d656ec3d6ee44ad2b405298e58f Support opendnssec 2.1.6
* c2e355ae59b8ecb862838eeeb37a33370a4874a4 Remove the <Interval> from opendnssec conf
* b857828180c650d7b2db17663a9413323eaef73d With opendnssec 2, read the zone list from file
* 8080bf7b359ca87d62502acf1fbdf235376007f7 Support OpenDNSSEC 2.1: new ods-signer protocol
* b6865831c97bbab71c02ec12dea2280caa505c6d DnsSecMaster migration: move the call to zonelist export later
* 682b59c8e801f06b3c6e4aa97c1ab33390c9edf8 opendnssec2.1 support: move all ods tasks to specific file

**Pull request closed without merging**